### PR TITLE
Avoid holding map mutex accross iterateRetryResources()

### PR DESCRIPTION
**Pull request by cathy-zhou**
during the scale testing we observed that the ovnkube-master
was not annotating the pods fast enough as a result the ovn-k8s-cni
on the scheduled node was timing out while waiting on the pod's
annotation. the pods were getting annotated by ovnkube-master after
several minutes.

this is because the iterateRetryResources() function was walking through
the list of pods that failed setup and was retrying setup on them with
lock held. there were 100s of these pods and each one of the took
several seconds.

The fix here is to retrieve the list of all the pods that needs to be
retried with retryPods map lock held and then call ensurePod() on
each of those pods with the new mutex in retryEntry struct held. This
will ensure that the add/update for Pod doesn't race with delete for
Pod.

Furthermore, in the iterateRetryResources() the call to handle resources
will be wrapped in a goroutine to allow multiple processing to execute in
parallel. This should result in iterateRetryResources() to complete its
task soon. A sync.waitgroup is used to wait for all the go routines to
complete.

Signed-off-by: Girish Moodalbail <gmoodalbail@nvidia.com>
Signed-off-by: Yun Zhou <yunz@nvidia.com>

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/ovn/obj_retry.go [hunk 1]
```diff
@@ -1050,102 +1090,145 @@ func (oc *Controller) deleteResource(objectsToRetry *retryObjs, obj, cachedObj i
...
+	}
+	r.retryMutex.Unlock()
+
+	// Now process the above list of pods that need re-try by holding the lock for each one of them.
+	klog.V(5).Infof("Going to retry resource setup for %d number of resource", len(localRetryEntries))
+
+	wg := &sync.WaitGroup{}
+	for _, lre := range localRetryEntries {
+		wg.Add(1)
+		go func(lre *localRetryEntry) {
```
Comment (msherif1234) [inline go-controller/pkg/ovn/obj_retry.go:1135]: can u move this function in separate function to make the code more readable

File: go-controller/pkg/ovn/obj_retry.go [hunk 2]
```diff
@@ -98,104 +99,141 @@ func (r *retryObjs) addRetryObj(obj interface{}) {
...
-	if entry, ok := r.entries[key]; ok {
-		entry.timeStamp = time.Now()
-		entry.newObj = obj
-	} else {
-		r.entries[key] = &retryObjEntry{newObj: obj,
-			timeStamp: time.Now(), backoffSec: 1, ignore: true}
-	}
+	entry := r.ensureRetryEntryLocked(key, &retryObjEntry{Mutex: sync.Mutex{}, newObj: obj,
+		timeStamp: time.Now(), backoffSec: 1, ignore: true})
+	entry.timeStamp = time.Now()
```
[the remark this answers is not in the document]
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:104]: Note that when the 2nd argument (newRetryEntry) of the ensureRetryEntryLocked() function is non-nil, the return value will not be nil.

File: go-controller/pkg/ovn/obj_retry.go [hunk 3]
```diff
@@ -98,104 +99,141 @@ func (r *retryObjs) addRetryObj(obj interface{}) {
...
+// step on each other. if newRetryEntry is specified, then the function will try to create a
+// new retryEntry in the map, otherwise it will return nil.
+func (r *retryObjs) ensureRetryEntryLocked(key string, newRetryEntry *retryObjEntry) *retryObjEntry {
 	r.retryMutex.Lock()
-	defer r.retryMutex.Unlock()
-	_, ok := r.entries[key]
-	return ok
+	entry := r.entries[key]
+	if entry != nil {
+		// do not hold on retryPodsLock while waiting on retryEntry to Lock
```
[the remark this answers is not in the document]
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:209]: fixed

File: go-controller/pkg/ovn/obj_retry.go [hunk 4]
```diff
@@ -98,104 +99,141 @@ func (r *retryObjs) addRetryObj(obj interface{}) {
...
-		}
-	} else {
-		r.entries[key] = &retryObjEntry{oldObj: obj, config: config,
-			timeStamp: time.Now(), backoffSec: 1, ignore: true}
+func (r *retryObjs) initRetryObjWithDelete(obj interface{}, key string, config interface{}, noRetryAdd bool) {
+	entry := r.ensureRetryEntryLocked(key, &retryObjEntry{Mutex: sync.Mutex{}, oldObj: obj, config: config,
+		timeStamp: time.Now(), backoffSec: 1, ignore: true})
+	entry.timeStamp = time.Now()
+	entry.oldObj = obj
+	entry.ignore = true
```
[the remark this answers is not in the document]
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:127]: I found not all initRetryObjWithDelete() functions are called following a skipRetryObj() function call. Is that a bug. But I guess this fix is not to address that even it is a bug. will remove.

File: go-controller/pkg/ovn/obj_retry.go [hunk 5]
```diff
@@ -98,104 +99,141 @@ func (r *retryObjs) addRetryObj(obj interface{}) {
...
 func (r *retryObjs) deleteRetryObj(key string, withLock bool) {
 	if withLock {
-		r.retryMutex.Lock()
-		defer r.retryMutex.Unlock()
+		entry := r.ensureRetryEntryLocked(key, nil)
+		if entry != nil {
+			defer entry.Unlock()
+		}
 	}
+	r.retryMutex.Lock()
```
[the remark this answers is not in the document]
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:177]: yes, that is intentional, withLock is only intent to indicate whether to lock hold an entry lock.

File: go-controller/pkg/ovn/obj_retry.go [hunk 6]
```diff
@@ -211,11 +249,17 @@ func (r *retryObjs) requestRetryObjs() {
...
 //getObjRetryEntry returns a copy of an object  retry entry from the cache
 func (r *retryObjs) getObjRetryEntry(key string) *retryObjEntry {
-	r.retryMutex.Lock()
-	defer r.retryMutex.Unlock()
-	if entry, ok := r.entries[key]; ok {
-		x := *entry
-		return &x
+	entry := r.ensureRetryEntryLocked(key, nil)
+	if entry != nil {
+		defer entry.Unlock()
```
[the remark this answers is not in the document]
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:254]: sure

File: go-controller/pkg/ovn/obj_retry.go [hunk 7]
```diff
@@ -773,17 +817,13 @@ func (oc *Controller) addResource(objectsToRetry *retryObjs, obj interface{}, fr
 
 // Given a *retryObjs instance, an old and a new object, updateResource updates the specified object in the cluster
 // to its version in newObj according to its type and returns the error, if any, yielded during the object update.
```
[the remark this answers is not in the document]
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:819]: sure

File: go-controller/pkg/ovn/obj_retry.go [hunk 8]
```diff
@@ -98,104 +99,141 @@ func (r *retryObjs) addRetryObj(obj interface{}) {
...
+	entry := r.ensureRetryEntryLocked(key, &retryObjEntry{Mutex: sync.Mutex{}, newObj: newObj, config: oldObj,
+		timeStamp: time.Now(), backoffSec: 1, ignore: true})
+	entry.timeStamp = time.Now()
+	entry.newObj = newObj
+	entry.config = oldObj
+	entry.Unlock()
 }
 
 // initRetryWithDelete tracks an object that failed to be deleted to potentially retry later
 // initially it is marked as skipped for retry loop (ignore = true)
```
[the remark this answers is not in the document]
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:121]: sure

File: go-controller/pkg/ovn/obj_retry.go (hunk 1 above)
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:1135]: will do. thanks very much for your comments. will make changes and repush.

Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:1231]: @msherif1234 repushed. please take a look. thanks

File: go-controller/pkg/ovn/obj_retry.go [hunk 9]
```diff
@@ -99,104 +100,140 @@ func (r *retryObjs) addRetryObj(obj interface{}) {
...
+		entry.Unlock()
+		return true
+	}
+	return false
+}
+
+// returns the retryEntry structure with lock held on it. all users of retryEntry should call
+// this function. the mutex and the `ignore` field ensures that the pod add/update/delete do not
+// step on each other. if newRetryEntry is specified, then the function will try to create a
+// new retryEntry in the map, otherwise it will return nil.
```
Comment (ricky-rav) [inline go-controller/pkg/ovn/obj_retry.go:205]: NIT: I would make the comment even more clear:
_if newRetryEntry is specified and the map has no entry with the input key, then the function will try to add newRetryEntry to the map_

File: go-controller/pkg/ovn/obj_retry.go [hunk 10]
```diff
@@ -1047,106 +1088,154 @@ func (oc *Controller) deleteResource(objectsToRetry *retryObjs, obj, cachedObj i
...
 }
 
+type localRetryEntry struct {
+	key string // the key in the retryObjs map holding retryObjs value
+	// cached newObj holds k8s resource failed during add operation
+	newObj interface{}
+	// cached oldObj holds k8s resource failed during delete operation
+	oldObj interface{}
+	// resource retrieved from the K8s API Server
+	kObj interface{} // Pod retrieved from the K8s API Server
```
Comment (ricky-rav) [inline go-controller/pkg/ovn/obj_retry.go:1098]: In the comment I'd replace `Pod` with `Resource`

File: go-controller/pkg/ovn/obj_retry.go [hunk 11]
```diff
@@ -1047,106 +1088,154 @@ func (oc *Controller) deleteResource(objectsToRetry *retryObjs, obj, cachedObj i
...
+			klog.Infof("%v retry: removing old object for %s", r.oType, objKey)
+			if !isResourceScheduled(r.oType, entry.oldObj) {
+				klog.V(5).Infof("Retry: %s %s not scheduled", r.oType, objKey)
+				return
+			}
+			if err := oc.deleteResource(r, entry.oldObj, entry.config); err != nil {
+				klog.Infof("%v retry delete failed for %s, will try again later: %v", r.oType, objKey, err)
+				r.retryMutex.Lock()
+				entry.timeStamp = time.Now()
+				r.entries[lre.key] = entry
```
Comment (ricky-rav) [inline go-controller/pkg/ovn/obj_retry.go:1159]: `lre.key` is `objKey`, right?

File: go-controller/pkg/ovn/obj_retry.go [hunk 12]
```diff
@@ -1047,106 +1088,154 @@ func (oc *Controller) deleteResource(objectsToRetry *retryObjs, obj, cachedObj i
...
+			klog.Infof("%v retry: creating object for %s", r.oType, objKey)
+			if !isResourceScheduled(r.oType, entry.newObj) {
+				klog.V(5).Infof("Retry: %s %s not scheduled", r.oType, objKey)
+				return
+			}
+			if err := oc.addResource(r, entry.newObj, true); err != nil {
+				klog.Infof("%v retry create failed for %s, will try again later: %v", r.oType, objKey, err)
+				r.retryMutex.Lock()
+				entry.timeStamp = time.Now()
+				r.entries[lre.key] = entry
```
Comment (ricky-rav) [inline go-controller/pkg/ovn/obj_retry.go:1178]: `objKey`

File: go-controller/pkg/ovn/obj_retry.go (hunk 9 above)
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:205]: accept

File: go-controller/pkg/ovn/obj_retry.go (hunk 11 above)
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:1159]: accept

File: go-controller/pkg/ovn/obj_retry.go [hunk 13]
```diff
@@ -1047,106 +1088,154 @@ func (oc *Controller) deleteResource(objectsToRetry *retryObjs, obj, cachedObj i
...
+	if entry.newObj != nil {
+		initObj = entry.newObj
+	} else if entry.oldObj != nil {
+		initObj = entry.oldObj
+	}
+
+	klog.Infof("%v %s: retry object setup", r.oType, objKey)
+
+	if entry.newObj != nil {
+		entry.newObj = lre.kObj
```
[the remark this answers is not in the document]
Comment (cathy-zhou) [reply go-controller/pkg/ovn/obj_retry.go:1135]: in the current code, entry.newObj was only set to the kObj, if entry.newObj is not nil. see https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/obj_retry.go#L1096.

[the remark this answers is not in the document]
Comment (ricky-rav) [reply go-controller/pkg/ovn/obj_retry.go:1165]: True, my bad! Let's leave it as it is.

**Comment by msherif1234** [comment]
LGTM @trozet PTAL

File: go-controller/pkg/ovn/obj_retry.go [hunk 14]
```diff
@@ -141,65 +134,110 @@ func (r *retryObjs) initRetryObjWithUpdate(oldObj, newObj interface{}, key strin
...
+		entry.Unlock()
+		return true
+	}
+	return false
+}
+
+// returns the retryEntry structure with lock held on it. all users of retryEntry should call
+// this function. the mutex and the `ignore` field ensures that the pod add/update/delete do not
+// step on each other. if the map has no entry with the input key and if newRetryEntry is specified,
+// then the function will try to create a new retryEntry in the map if , otherwise it will return nil.
```
Comment (ricky-rav) [inline go-controller/pkg/ovn/obj_retry.go:209]: there's an extra "if " before the last comma :)

**Comment by ricky-rav** [comment]
/LGTM 
Thanks, @cathy-zhou!

File: go-controller/pkg/ovn/obj_retry.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/ovn/obj_retry.go:1165]: is this a regression? I thought we always used to get the latest object for pods at least: https://github.com/openshift/ovn-kubernetes/blob/release-4.10/go-controller/pkg/ovn/pods_retry.go#L37

**Pull request merged**